Doc 1448 please add spark sql direct execution feature to of as documentation (#3627)
spotinst/help: src/docs/spark-connect/interactive/README.md

> # Ofas Interactive Applications

Reviewer: ```suggestion
# Ocean Spark Interactive Applications
```

## Introduction

> An interactive Spark application is a long-running application which the user can interact with on runtime. This can be a notebook application or applications running Spark Connect or HiveThrift servers. Spark Connect supports languages such as Python, Scala and Go and potentially all languages supported by gRPC. The HiveThrift server offers SQL interaction through JDBC.

Reviewer: ```suggestion
An interactive Spark application is a long-running application which the user can interact with on runtime. This can be a notebook application or applications running [Spark Connect](https://spark.apache.org/docs/latest/spark-connect-overview.html) or a [Thrift Hive Server](https://cwiki.apache.org/confluence/display/hive/hiveserver). Spark Connect supports languages such as Python, Scala and Go and potentially all languages supported by [gRPC](https://grpc.io). The Thrift Hive Server offers SQL interaction through JDBC.
```

## Spark Connect

> In Apache Spark 3.4, Spark Connect introduced a decoupled client-server architecture that allows remote connectivity to Spark clusters using the DataFrame API and unresolved logical plans as the protocol. The separation between client and server allows Spark and its open ecosystem to be leveraged from everywhere. It can be embedded in modern data applications, in IDEs, Notebooks and programming languages.

Reviewer: ```suggestion
In Apache Spark 3.4, Spark Connect introduced a decoupled client-server architecture that allows remote connectivity to Spark clusters using the DataFrame API and unresolved logical plans as the protocol. The separation between client and server allows Spark and its open ecosystem to be leveraged from everywhere. It can be embedded in modern data applications, in IDEs, Notebooks and programming languages. Ocean Spark support Spark Connect, and that can be especially useful for the direct execution of Spark SQL.
```

> 

Reviewer: @sigmarkarl Can you please add some lines here explaining direct execution of Spark SQL and give an example for customer to run?

Author: added Using dbt section at the botton, following JDBC

```python
from ocean_spark_connect.ocean_spark_session import OceanSparkSession

spark = OceanSparkSession.Builder().cluster_id("osc-cluster").appid("appid").profile("default").getOrCreate()
spark.sql("select random()").show()
spark.stop()
```

#### Spotctl

Use the spotctl command line tool to open a websocket proxy to the interactive Spark application

```sh
brew install spotinst/tap/spotctl
spotctl configure
```

```sh
spotctl ocean spark --clusterid osc-cluster --appid appid --profile default
```

spotctl will start a service on port 15002 (the default Spark Connect port)

```sh
pyspark --remote sc://localhost
```

### Example

Start the application using Postman, from the console or the command line

```sh
curl -k -X POST 'https://api.spotinst.io/ocean/spark/cluster/{clusterId}/app?accountId={accountId}' -H 'Content-Type: application/json' -H 'Authorization: Bearer {token}' -d '
```

```json
{
  "jobId": "spark-connect",
  "configOverrides": {
    "mainClass": "org.apache.spark.connect.sql.service.SparkConnectServer",
    "deps": {
      "packages": ["org.apache.spark:spark-connect_2.12:3.5.0"]
    }
  }
}
```

> 

Reviewer: @thorsteinnth @sigmarkarl  Can you please add an example here of how to use https://www.getdbt.com with Spark Connect? You could hyperlink to the existing part of our documentation that describes how to use AWS Glue Catalog and Hive Metastore.

Author: Added a Using dbt section at the bottom, following the JDBC discussion

## JDBC

> JDBC, or Java Database Connectivity, is an API used in Java programming to interact with databases. It provides a standard abstraction for Java applications to communicate with various databases1. JDBC allows applications to send requests made by users to the specified database1. It is used to write programs required to access databases. Apache Spark provides a JDBC interface throught the HiveThriftServer.

Reviewer: ```suggestion
JDBC, or Java Database Connectivity, is an API used in Java programming to interact with databases. It provides a standard abstraction for Java applications to communicate with various databases1. JDBC allows applications to send requests made by users to the specified database1. It is used to write programs required to access databases. Apache Spark provides a JDBC interface through the HiveThriftServer.
```

### Server side

To enable JDBC connections to the Spark Application, start the HiveThriftServer

#### Launch a JDBC server

```json
"mainClass": "com.netapp.spark.HiveThriftServer",
"deps": {
    "packages": ["com.netapp.spark:hive:1.2.0"],
    "repositories": ["https://us-central1-maven.pkg.dev/ocean-spark/spark-code-submission-plugin"]
}
```

### Client side

> 

Reviewer: @sigmarkarl @thorsteinnth Can you please add here an example of how to launch a Spark Streaming application, even if it is only a very brief example.

> #### Ofas JDBC driver

Reviewer: ```suggestion
#### Ocean Spark JDBC driver
```

> Use the Ofas JDBC driver with database tool or in your code project. The driver available at the following maven coordinates

Reviewer: ```suggestion
Use the Ocean Spark JDBC driver with database tool or in your code project. The driver available at the following maven coordinates
```

Once connected to the remote application (as described in the Client side section below) the user can execute SQL queries directly from code, notebook or pyspark shell

```Python
df = spark.sql("select 'apple' as word, 123 as count union all select 'orange' as word, 456 as count")
df.write.save("s3://results_bucket/fruits.parquet")
```

### Server side

To start a Spark application with SparkConnect server, either run the mainClass SparkConnectServer or enable the SparkConnect plugin. Using the Spark Connect plugin, the application can run other tasks or services while enabling Spark Connect.

#### Spark Connect launch using the SparkConnectServer

```json
"mainClass": "org.apache.spark.connect.sql.service.SparkConnectServer",
"deps": {

>     "packages": ["org.apache.spark:spark-connect_2.12:3.5.0"]

Reviewer 2: should this be 3.4?

```python
from ocean_spark_connect.ocean_spark_session import OceanSparkSession

spark = OceanSparkSession.Builder().cluster_id("osc-cluster").appid("appid").profile("default").getOrCreate()
spark.sql("select random()").show()
spark.stop()
```

#### Spotctl

Use the spotctl command line tool to open a websocket proxy to the interactive Spark application

```sh
brew install spotinst/tap/spotctl
spotctl configure
```

```sh
spotctl ocean spark --clusterid osc-cluster --appid appid --profile default
```

spotctl will start a service on port 15002 (the default Spark Connect port)

```sh
pyspark --remote sc://localhost
```

### Example

Start the application using Postman, from the console or the command line

```sh
curl -k -X POST 'https://api.spotinst.io/ocean/spark/cluster/{clusterId}/app?accountId={accountId}' -H 'Content-Type: application/json' -H 'Authorization: Bearer {token}' -d '
```

```json
{
  "jobId": "spark-connect",
  "configOverrides": {
    "mainClass": "org.apache.spark.connect.sql.service.SparkConnectServer",
    "deps": {
      "packages": ["org.apache.spark:spark-connect_2.12:3.5.0"]
    }
  }
}
```

## JDBC

> JDBC, or Java Database Connectivity, is an API used in Java programming to interact with databases. It provides a standard abstraction for Java applications to communicate with various databases1. JDBC allows applications to send requests made by users to the specified database1. It is used to write programs required to access databases. Apache Spark provides a JDBC interface through the HiveThriftServer.

Reviewer 2: databases1 -> databases

The user can execute SQL queries directly by using the JDBC driver in code, a database tool or from another Spark session. Here is a Java example

```Java
var prop = new Properties();
var query = "select 'apple' as word, 123 as count union all select 'orange' as word, 456 as count";
var jdbcUrl = "jdbc:ofas://api.spotinst.io/"+clusterId+"/"+appId+"?profile=default";
//var hiveUrl = "jdbc:hive2://localhost:10000";

try (var conn = DriverManager.getConnection(jdbcUrl, prop); var stmt = conn.createStatement(); var rs = stmt.executeQuery(query)) {
  var metadata = rs.getMetaData();
  int columnCount = metadata.getColumnCount();
  while(rs.next()) {
      var row = new StringBuilder();
      int i;
      for (i = 1; i < columnCount; i++) {
          row.append(rs.getString(i)).append(", ");
      }
      row.append(rs.getString(i));
      System.out.println(row);
  }
}
```

### Server side

To enable JDBC connections to the Spark Application, start the HiveThriftServer

#### Launch a JDBC server

```json
"mainClass": "com.netapp.spark.HiveThriftServer",
"deps": {
    "packages": ["com.netapp.spark:hive:1.2.0"],

>     "repositories": ["https://us-central1-maven.pkg.dev/ocean-spark/spark-code-submission-plugin"]

Reviewer 2: we should move this under a new repo and maven coordinates
